AlphaBetaGame.java

```diff
@@ -1,149 +1,99 @@
-			int bestMoveY, bestMoveX;
 
-			public ValueWithMove(int outcome, int lengthOfGame, int bestMoveY, int bestMoveX) {
-				super(outcome, lengthOfGame);
-				this.bestMoveY = bestMoveY;
-				this.bestMoveX = bestMoveX;
-			}
-	}
-	@override
-	protected int computerMoves() {
-		/* Computer moves. Afterwards returns
-		0 : Nobody won yet
-		2 : Computer won
-		3 : Table is full  */
-		
-		ValueWithMove move = alphaBeta(table, 2, new Value(-1, 0), new Value(1,0));
-		System.out.println("Outcome: " + move.outcome + " Length: " + move.lengthOfGame);
-		(new Scanner(System.in)).nextLine();
-		table[move.bestMoveY][move.bestMoveX] = 2;
-		if (whoWon(move.bestMoveY, move.bestMoveX, 2) == 2)
-			return 2;
-		else if (isFull())
-			return 3;
-		else
-			return 0;
+		// If it needs to be computed
+		ValueWithMove bestMove = alphaBeta(new Value(-1, 0), new Value(1,0));
+		dataBase.put(key, new int[] {bestMove.outcome, bestMove.lengthOfGame,
+									bestMove.bestMoveY, bestMove.bestMoveX});
+		return bestMove;
 	}
-	protected ValueWithMove alphaBeta(int[][] table, int player, Value alpha, Value beta) { 
-		// RETURNS: 
-		//    First entry:  The value of the position for player when it is his move.
-		//    Second entry: The length of the resulting game using optimal play.
-		//	  Third entry:  Optimal move Y
-		//	  Fourth entry: Optimal move X
-		// 
-		// NOTE: opponent = 3 - player. The second coordinate is needed so that the 
-		// computer player wants to win as quickly as possible or survive as long 
-		// as it can.
-
-		// Find out if game is over.
-		int whoWon = 0;
-
-		for(int i = 0; i < table.length; i++) {
-			for(int j = 0; j < table[i].length; j++) {
-				if(table[i][j] == 3 - player) {
-					if (Game.whoWon(table, i, j, 3-player, lengthToWin) == 3-player)
-						whoWon = 3 - player;
-				}
-				if(table[i][j] == player) {
-					if (Game.whoWon(table, i, j, player, lengthToWin) == player)
-						whoWon = player;
-				}
-			}
-		}
-
+	protected ValueWithMove alphaBeta(Value alpha, Value beta) { 
+		// Getting best move at a current state of the table.
+		// RETURNS: Value of position and best move realizing that value.
+		//    
+		// NOTE: opponent = 3 - player.
+		//
 		// Return the value if the game is over after opponents move.
-		// In this case (-1,-1) for best move in place of null.
-		// These final states are only visited during the recursion.
-		if (whoWon == player) {
-			// printTable();
-			// System.out.println("Player won: " + player);
-			// (new Scanner(System.in)).nextLine();
-			return new ValueWithMove(1, 0, -1, -1);
-		}
-		else if (whoWon == 3-player) {
-			// printTable();
-			// System.out.println("Player won: " + (3-player));
-			// (new Scanner(System.in)).nextLine();
-			return new ValueWithMove(-1, 0, -1, -1);
+		// In this case (-1,-1) in place of null.
+
+		int tempAgent = activeAgent;
+		int tempY = lastMoveY;
+		int tempX = lastMoveX;
+		
+		activeAgent = 3 - tempAgent;
```

Comment:
It is because of the next line when I call winOnLastMove. Since this is the player's turn, only the computer could have won on the last move. I know bad design...